### PR TITLE
Batch Size Rampup

### PR DESCRIPTION
**Pull request by kuba-krj**
# Description
todo

# Checklist
   - [ ] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [ ] Are all variables set with logical defaults that are backward compatible?
   - [ ] I attached link to neptune if it was necessary
   - [ ] Have you successfully executed the linter and tests?
   - [ ] Are all functions concise and don't require splitting?
   - [ ] Is there documentation for arguments and complex logic?
   - [ ] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [ ] Are all functions placed in appropriate files?
   - [ ] Is the PR name meaningful and descriptive?
   - [ ] Is PR description provided, or unnecessary?
  
Other questions:
   - [ ] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [ ] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [ ] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

**Comment by j321m** [comment]
If possible add an experiment showing that higher batch size with rampup has same loss as lower batch size, corresponding to the rampup value

File: research/conditional/utils/conditional_trainer.py [hunk 1]
```diff
@@ -190,20 +192,42 @@ def _train_step(
...
+                num_processed_tokens,
+                self.batch_size_rampup_transition_points,
+                self.batch_size_rampup_sizes,
+                self.batch_size,
+            )
+        num_bs_chunks_from_rampup = self.batch_size // current_bs
+
+        for i in range(num_bs_chunks_from_rampup):
+            # TODO: make a way to avoid copying the whole batch just to get a slice
+            batch_copy = copy.deepcopy(processed_batch)
```
[the remark this answers is not in the document]
Comment (kuba-krj) [reply research/conditional/utils/conditional_trainer.py:209]: let's meet and discuss how to best refactor it

**Comment by j321m** [comment]
there seem to be something wrong with logging to neptune after the changes (1st neptune experiments are ok, 2nd aren't). The issue is, that loss curves of loss/step are different to loss curves of loss/token_per_active (loss at 100 for example)

**Review by crewtool — CHANGES_REQUESTED** [review]
Left some comments

File: lizrd/support/misc.py [hunk 2]
```diff
@@ -8,6 +8,11 @@ def tags_to_name(tags: Optional[List[str]]) -> str:
     return "_".join(tags) if tags else ""
```
Comment (j321m) [inline lizrd/support/misc.py:10]: there are functions like that, but no big deal i think

File: lizrd/support/misc.py [hunk 3]
```diff
@@ -66,3 +71,21 @@ def get_ith_chunk(tensor, chunks, i):
 
     list_of_chunks = torch.chunk(tensor, chunks, dim=0)
     return list_of_chunks[i]
+
+
+def calculate_current_bsz_from_rampup(
```
[the remark this answers is not in the document]
Comment (kuba-krj) [reply lizrd/support/misc.py]: I agree, changed

File: research/datasets.py [hunk 4]
```diff
@@ -1,19 +1,66 @@
...
 
 import torch
 from torch.utils.data import DataLoader
 
 from lizrd.text import datasets, packers, data, tokenizers
+from lizrd.support.misc import get_ith_chunk
+
+
+@dataclass
+class BatchSizeRampupConfig:
```
Comment (j321m) [inline research/datasets.py:14]: can we have some assert that will make sure that this class stores 2 lists of ints of the same length?

Comment (kuba-krj) [reply research/datasets.py:14]: We have this assert in check_args. I am not sure if it makes sense to add it also here

File: research/conditional/utils/conditional_trainer.py [hunk 5]
```diff
@@ -370,11 +404,12 @@ def _decode_samples(self, step):
                 iteration=step,
             )
```
Comment (j321m) [inline research/conditional/utils/conditional_trainer.py:405]: I don't like abbreviating `batch_size` to `bsz`

Comment (kuba-krj) [reply research/conditional/utils/conditional_trainer.py:405]: ok, I changed it

File: research/datasets.py (hunk 4 above)
Comment (j321m) [reply research/datasets.py:14]: it would be for code readability.

Comment (j321m) [reply research/datasets.py:14]: I'd also add 1 line description for the class, about what is it for, since config is very widely used sting in this repo

**Comment by j321m** [comment]
I don't see any mistakes by looking at the code. Please add neptune tests and I can give my approve.
test suggestion:
same model, 3 runs:
1st batch_size = 512.
2nd batch_size = 2048, but rampup is 512 and lasts for the whole training. loss schould be similar to 1st (like with different data_seed).
3rd batch size rampup utilization, so that loss is better than in 1 and 2.

also, last time neptune logging with rampup was [funny](https://app.neptune.ai/o/pmtest/org/llm-random/runs/compare?viewId=9a15aa36-1063-49c4-807d-53393abb118a&dash=charts&compare=IwTgbGAcA0xA&compareChartsFilter=loss_interval%2F1000&query=(sys%2Fid%3Astring%20%3D%20%22LLMRANDOM-19669%22)%20OR%20(sys%2Fid%3Astring%20%3D%20%22LLMRANDOM-19668%22)&sortBy=%5B%22sys%2Fcreation_time%22%5D&sortFieldType=%5B%22datetime%22%5D&sortFieldAggregationMode=%5B%22auto%22%5D&sortDirection=%5B%22descending%22%5D&suggestionsEnabled=false&lbViewUnpacked=true). make sure that doesn't happen.

File: lizrd/support/test_misc.py
Comment (crewtool) [inline lizrd/support/test_misc.py:1]: nice!

File: lizrd/support/misc.py [hunk 6]
```diff
@@ -224,3 +229,21 @@ def get_ith_chunk(tensor, chunks, i):
...
+    transition_points_in_tokens = [point * 1e9 for point in transition_points]
+
+    if processed_tokens >= transition_points_in_tokens[-1]:
+        return target_batch_size
+
+    for i in reversed(range(len(transition_points_in_tokens) - 1)):
+        if processed_tokens >= transition_points_in_tokens[i]:
+            return batch_sizes[i + 1]
+
+    return batch_sizes[0]
```
Comment (crewtool) [inline lizrd/support/misc.py:249]: Just a suggestion. Suppose our object BatchSizeRampupConfig instead is a list of pairs (transition_point, batch_size). We can simplify a function : 
```suggestion
def calculate_current_batch_size_from_rampup(
    processed_tokens,
    rampup_transition_config: List[int, int], 
    target_batch_size,
):
    for transition_point, batch_size in rampup_transition_config:
        if processed_tokens < transition_point * 1e9:
            return batch_size
    return target_batch_size
```

File: research/datasets.py (hunk 4 above)
Comment (crewtool) [inline research/datasets.py:14]: My initial suggestion was to replace two list with a list of tuples. 
For  transition_points = [0.5, 1, 2]  and batch_sizes = [16, 64, 128] representing
```math
f(processedtokens) = 
     \begin{cases}
       \text{16} &\quad\text{if} 0 \leq \text{processedtokens} \le 0.5 \\
       \text{64} &\quad\text{if} 0.5 \leq \text{processedtokens} \le 1 \\
       \text{128} &\quad\text{if} 1 \leq \text{processedtokens} \le 2 \\
       \text{512} &\quad \text{otherwise}
     \end{cases}
```

Sorry if I wasn't clear.

**Review by crewtool — CHANGES_REQUESTED** [review]
I disagree changing logger. 
Other things are just suggestions.

File: lizrd/support/misc.py (hunk 6 above)
Comment (crewtool) [reply lizrd/support/misc.py:249]: We can also add 
```math
f(processedtokens) = 
     \begin{cases}
       \text{16} &\quad\text{if} 0 \leq \text{processedtokens} \le 0.5 \\
       \text{64} &\quad\text{if} 0.5 \leq \text{processedtokens} \le 1 \\
       \text{128} &\quad\text{if} 1 \leq \text{processedtokens} \le 2 \\
       \text{512} &\quad \text{otherwise}
     \end{cases}
```

File: research/conditional/utils/conditional_trainer.py [hunk 7]
```diff
@@ -384,7 +421,7 @@ def _log_train_stats(self, loss_value, step):
                     title=name,
                     value=stats.acc / stats.interval,
                     iteration=step,
-                    processed_token_scale=True,
+                    processed_tokens_so_far=self.num_processed_tokens,
```
Comment (crewtool) [inline research/conditional/utils/conditional_trainer.py:424]: Since we cannot calculate processed_tokens based on iteration,I think we should change code this way:
```suggestion
                self.logger.report_scalar(
                    title=name,
                    value=stats.acc / stats.interval,
                    iteration=step,
                 )
                 self.logger.report_scalar(
                    title=name,
                    value=stats.acc / stats.interval,
                    iteration= processed_tokens_so_far,
                 )
            
```

File: lizrd/support/logging.py [hunk 8]
```diff
@@ -342,12 +346,15 @@ def report_scalar(
...
+        processed_tokens_so_far: Optional[int] = None,
     ):
         path = self._make_path(title, series)
         wandb.log({path: value, "train/step": iteration})
         auxiliary_metrics = self.get_auxiliary_metrics(
-            title, value, iteration, token_scale=processed_token_scale
+            title,
+            value,
+            iteration,
+            processed_tokens_so_far=processed_tokens_so_far,
```
Comment (crewtool) [inline lizrd/support/logging.py:357]: We should delete token_scale / processed_tokens_so_far as it is no longer relevant

File: lizrd/support/logging.py [hunk 9]
```diff
@@ -150,13 +150,17 @@ def with_token_scale(self, title: str, value: float, iteration: int):
...
+        iteration: int,
+        processed_tokens_so_far: Optional[int],
     ):
         auxiliary_metrics = {}
-        if token_scale:
+        if processed_tokens_so_far is not None:
             auxiliary_metrics = {
                 **auxiliary_metrics,
-                **self.with_token_scale(title, value, iteration),
+                **self.with_token_scale(title, value, processed_tokens_so_far),
```
Comment (crewtool) [inline lizrd/support/logging.py:163]: I think it would be better to delete it.

File: lizrd/support/logging.py [hunk 10]
```diff
@@ -150,13 +150,17 @@ def with_token_scale(self, title: str, value: float, iteration: int):
...
             }
```
Comment (crewtool) [inline lizrd/support/logging.py:164]: i think we should remove this part.

File: lizrd/support/logging.py [hunk 11]
```diff
@@ -133,15 +134,14 @@ def get_metric_with_flop_scale(self, value: float, iteration: int):
             * self.args["batch_size"],
         }
 
-    def with_token_scale(self, title: str, value: float, iteration: int):
-        tokens_passed = iteration * self.args.get("tokens_per_step")
-        tokens_per_active_params = tokens_passed / self.args.get(
+    def with_token_scale(self, title: str, value: float, processed_tokens_so_far: int):
```
Comment (crewtool) [inline lizrd/support/logging.py:137]: I think we should remove it.

**Review by crewtool — APPROVED** [review]
LGTM. Please consider comments.

File: lizrd/support/misc.py (hunk 6 above)
Comment (kuba-krj) [reply lizrd/support/misc.py:249]: Right! I will change it as suggested by you

File: research/datasets.py (hunk 4 above)
Comment (kuba-krj) [reply research/datasets.py:14]: ok, adding

Comment (kuba-krj) [reply research/datasets.py:14]: I may have misunderstood, sorry. I think we can easily rethink/improve this part in the future
